feat(minifier): compress array of string literals to `'str1,str2'.split(',')`

sapphi-red opened the pull request:
Ported `["str1", "str2", ...]` => `"str1 str2".split(" ")` compression from closure compiler with some tweaks.

sapphi-red commented:
* **#8786** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8786?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/8786?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=8786).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red commented on crates/oxc_minifier/src/peephole/substitute_alternate_syntax.rs:
```diff
@@ -939,6 +940,75 @@ impl<'a> LatePeepholeOptimizations {
+        }
+
+        let element_count = array.elements.len();
+        // replace with `.split` only when the saved size is great enough
+        // because using `.split` in some places and not in others may cause gzipped size to be bigger
+        let can_save = element_count * 2 > ".split('.')".len() + THRESHOLD;
+        if !can_save {
+            return None;
+        }
```
closure compiler always replaces with `.split` when the saving amount is bigger than 0 (same as setting `THRESHOLD = 0` ). On one hand, that reduce the non-gzipped code size. But on the other hand, it increases the gzipped code size in many cases. To prevent that from happening, I introduced the `THRESHOLD` parameter. I manually selected the value `40` by checking the minsize output. This may be overfit to the benchmark, but I guess it's better than 0.

sapphi-red commented on crates/oxc_minifier/src/peephole/substitute_alternate_syntax.rs:
```diff
@@ -939,6 +940,78 @@ impl<'a> LatePeepholeOptimizations {
+    ) -> Option<&'static str> {
+        // These delimiters are chars that appears a lot in the program
+        // therefore probably have a small Huffman encoding.
+        const DELIMITERS: [&str; 5] = [" ", ".", ",", "(", ")"];
```
picked a different characters so that it matches our output

Boshen commented:
### Merge activity

* **Jan 30, 10:50 PM EST**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.
* **Jan 30, 10:50 PM EST**: A user added this pull request to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc).

sapphi-red committed:
feat(minifier): compress array of string literals to `'str1,str2'.split(',')` (#8786)

Ported `["str1", "str2", ...]` => `"str1 str2".split(" ")` compression from closure compiler with some tweaks.
